src/main/java/com/google/sps/servlets/NewsServlet.java

```diff
@@ -0,0 +1,114 @@
+    terms.forEach((term) -> {
+      try {
+        Map<String, String> paramMap = new HashMap<>();
+        paramMap.put("q", UrlRequest.encodeTerm(term));
+        paramMap.put("hl", "en-US");
+        paramMap.put("gl", "US");
+        paramMap.put("ceid", "US:en");
+        HttpURLConnection connect = getConnection(GOOGLE_NEWS_PATH, paramMap);
+        String HTMLString = getHTML(connect);
+        List<Article> articleList = makeArticleList(HTMLString);
+        articleMap.put(term, articleList);
+      }
+      catch (IOException e) {
+        e.printStackTrace();
+      }
+    });
+    String json = encodeMapAsJson(articleMap);
+    response.setCharacterEncoding("UTF-8");
+    response.getWriter().println(json);
+  }
+
+  public HttpURLConnection getConnection(String basePath, Map<String, String> paramMap) throws IOException {
+    String path = String.format("%s?%s", basePath, UrlRequest.getParamsString(paramMap));
+    URL url = new URL(path);
+    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+    connection.setRequestMethod("GET");
+    connection.setRequestProperty("Accept", "text/html");
+    connection.setRequestProperty("Content-Type", "text/html");
+    int responseCode = connection.getResponseCode();
+    if (responseCode != 200) {
+      System.err.println("Error: connection response code is: " + responseCode);
+    }
+    return connection;
+  }
+
+  public String getHTML(HttpURLConnection connection) throws IOException {
+    InputStreamReader inputStream = new InputStreamReader(connection.getInputStream());
+    StringBuilder contentBuilder = new StringBuilder();
+    try (BufferedReader inputReader = new BufferedReader(inputStream)) {
+      String inputLine;
+      while ((inputLine = inputReader.readLine()) != null) {
+        contentBuilder.append(inputLine);
+      }
+    } catch(IOException e) {
+      System.err.println(e);
+      return "";
+    } finally {
+      connection.disconnect();
+    }
+    return contentBuilder.toString();
+  }
+
+  public List<Article> makeArticleList(String HTMLString) throws IOException {
+    List<Article> articleList = new ArrayList<>();
+    String[] substrings = HTMLString.split(articleTag);
+    String thisSubstring, lastSubstring, title, link;
+    int closeBracket, openBracket;
+    // For each substring seperated by the article class tag.
+    for (int i = 5; i < substrings.length; i++) {
+      // Get the link from before the article tag and the title from after to make a new Article object.
+      lastSubstring = substrings[i - 1];
+      thisSubstring = substrings[i];
+      title = thisSubstring.substring(thisSubstring.indexOf(">") + 1, thisSubstring.indexOf("<"));
+      link = lastSubstring.substring(lastSubstring.lastIndexOf("href"), lastSubstring.lastIndexOf("\""));
+      link = "news.google.com" + link.substring(link.indexOf(".") + 1, link.lastIndexOf("\""));
+      articleList.add(new Article(title, link));
+    }
+    return articleList;
```

Comment:
[Optional] Use an API for web scrapping instead of doing it manually. 
[https://www.scrapingbee.com/blog/introduction-to-web-scraping-with-java/](https://www.scrapingbee.com/blog/introduction-to-web-scraping-with-java/)

Reply:
Andrew and I were thinking of putting some common methods in a util class after we both merged our original PR's. I think at that stage, it may be a good idea to transfer over to an API.

Reply:
Actually it looks like that API isn't free, so nvm.